### ISSUE TITLE
Messages are being skipped

### ISSUE THREAD

**Issue by bradurani** [zendesk/ruby-kafka#624]
* Version of Ruby: 2.5
* Version of Kafka: 0.11.0.2
* Version of ruby-kafka: 0.6.6

We're seeing behavior that only appears under production load that is causing messages to be skipped. Here's what we're seeing:

![image](https://user-images.githubusercontent.com/4195952/43544917-45a3197c-9589-11e8-8bc9-c5869b366a45.png)

Here you can see `consumer.message` for 2 consumer groups. As you can see the group A, represented by the top line, is consuming more message than group B, represented by the bottom line. This is despite the fact that they're consuming the same topic. Since one is clearly outpacing the other, you would expect to see lag in group B, however:

![image](https://user-images.githubusercontent.com/4195952/43545100-c8d72ab8-9589-11e8-8244-4e3375abc150.png)

Lag is actually ver small, fluctuating by only a few hundred messages per partition. At the same time, group B is consuming hundreds of thousands of message less every hour. For the same hour, here's
`producer.produce`
`consumer.messages` for A
`consumer.messages` for B
Sum of max `consumer.lag` for all partitions for B
![image](https://user-images.githubusercontent.com/4195952/43546235-f56c6f72-958c-11e8-8f61-05ab225e0c67.png)

As you can see there's a gap of 1324688 - 1088320 - 5616 = 230752 messages

Indeed, when we look at the destination db the records are being written to, the gap exists. We record partition and offset in the destination database, and there are gaps of hundreds of offsets consistent with the above graphs. Not only that, group A, which is the healthier group, also has gaps, though not as many. 
Both groups rarely fail to write to the destination, and when they do, we retry until they succeed before proceeding with the next batch.

I don't understand how we could have huge discrepancies in `consumer.messages` for 2 groups consuming the same topic, without seeing commensurate lag.

I haven't been able to reproduce this on my dev env. One thing I have noticed is that in production, both groups are logging `WARN -- : Reached max fetcher queue size (100), sleeping 1s`, and group B is showing this a lot more. 

Configurations I've tried:
- `automatically_mark_as_processed: true` with commit offsets every 10 seconds
- `automatically_mark_as_processed: false` with `mark_as_processed` and `commit_offsets` after every batch

Any help would be appreciated

**Comment by bradurani** [comment]
Update: tried with ruby-kafka: `0.6.7` and issue persists

**Comment by mensfeld** [comment]
@bradurani thanks for reporting.

Couple things: `Reached max fetcher queue size` means that there's a lot of data prefetched to the ruby-kafka. It may mean, you're falling behind with the processing and that your system produces more data than it can consume fast enough. If your retention policy is short enough it may cause weird behaviors that may look like losing data.

Could you try with a much smaller `fetcher_max_queue_size`? For example 2 and see if it persists or behaves differently? Could you try distributing the workload to more partitions?

**Comment by bradurani** [comment]
Updated: we added code to detect skipped messages:

```ruby
consumer.each_batch do |batch|

    check_for_skipped_offsets!(batch)

    process_batch(batch)
  end
end

def check_for_skipped_offsets!(batch)
  @previous_last_offsets ||= {}
  @previous_last_offsets[batch.topic] ||= {}
  previous_last_offset = @previous_last_offsets[batch.topic][batch.partition]

  if previous_last_offset.present? &&
    batch.messages.any? &&
    previous_last_offset != (batch.messages.first.offset - 1)
    num_skipped = batch.first_offset - previous_last_offset - 1
       log_discrepency(...)
    end
    @previous_last_offsets[batch.topic][batch.partition] = batch.last_offset
  end
end
```

Sure enough, the discrepancy becomes obvious right away:

![image](https://user-images.githubusercontent.com/4195952/43600094-85abd2c4-963e-11e8-9c73-271252bdec04.png)

**Comment by bradurani** [comment]
Update: we've set the `fetcher_max_queue_size` to 2, and if anything we're losing more messages

**Comment by bradurani** [comment]
Update: we switched to `master` and the issue persists

Update: We might have fixed it. We doubled the number of consumers and, while we still need to give it time to be sure, it looks like this might have fixed the issue.
It looks like the skipped messages occurred exactly when we had small amounts of lag. During busy periods, we had lag, but only a few hundred messages (see first post in the thread for graph). Is it possible there's a logical error around offset counting when we're not consuming at the head of the queue?
I'm left with a lot of lingering concerns. Since Kafka guarantees at-least-once delivery, it seems to me that we should never skip messages, even if our configuration is not ideal. I know there are a lot of variables to consider - lag, min bytes, max bytes, fetcher max queue size, number of partitions, number of consumers, Kafka version, compression - but even if we don't find the optimal consumer configuration, shouldn't the gem consume in a way that doesn't break the at-least-once delivery guarantees?
Or maybe we can work backwards: are there any configuration mistakes anyone out there knows of that are known to cause skipped messages?

**Comment by bradurani** [comment]
I have verified the missing messages in the downstream database and they are definitely being skipped. The missing message counts for for the downstream database match the numbers detected by my skipped message detection you see above. It's missing whole batches (10-200 records depending on load).
Topics: 1
Partitions: 32
Compression: GZIP
Producers ~100 servers for 1 app
I have 2 consumer groups consuming the same topic:
Consumers: 6 per group
Mostly Healthy Group:
  Consumer Loop Duration: 500ms
  Loss Rate: ~1%
Very Unhealthy Group:
  Consumer Loop Duration: 4,000ms
  Loss Rate: ~20%
max_bytes: 500,000
max_wait_time: 1
min_bytes: 1024 * 20
AutoCommit: every 10 seconds (I've also tried other configurations)

During times of low load (< 100 msg /sec). The problems goes away. At high load (500-600 msgs / sec), the problem is severe.

I fixed the problem by increasing the consumer count to 12 per group. Since then, I've lost only one batches of 32 messages that occurred during a rebalance.

Not sure if this is relevant, but during the times it's losing messages, partitions are alternating between 0 lag and small amount of lag (less than the size of a typical batch). Could it be a logical error calculating next batch to fetch? Would GZIP complicate that?

**Comment by nguyenquangminh0711** [comment]
@bradurani One more thing. Are all the producers using Ruby too? If no, could you please check whether other producers are using fancy features like Transactional or Idempotent messages production or not. I'm working on that feature support for ruby-kafka, and fixed a the bug of skipped messages because of transactional message production. Hope that would provide more information for you.

**Comment by bradurani** [comment]
W'e're using `ruby-kafka` as a producer and no, we're not using the features

**Comment by bradurani** [comment]
I'll try it tomorrow morning. We're using `0.11.0.2` BTW

**Comment by bradurani** [comment]
Update: we've deployed Nguyễn's bug fix branch and the problem stopped immediately

Also, we ported the program to Java with the exact same parameters and experienced zero data loss

@nguyenquangminh0711 your PR mentions the fix is for `newer versions of Kafka (after 0.11.0, I guess)`, but we're running `0.11.0.2` and were affected by the bug. Does this gel with your understanding?

**Comment by nguyenquangminh0711** [comment]
@bradurani Oh I mean this happens from Kafka 0.11.x, where the new troublesome record batch encoding format is introduced. So, your company's current version `0.11.0.2` is within the range I mentioned. Sorry for my bad English 😃. 

Btw, glad to hear the problem is gone. The PR is ready to merge. Perhaps we should have a bug fix release for 0.6.x beside the master branch.

**Comment by mensfeld** [comment]
@nguyenquangminh0711 
> Perhaps we should have a bug fix release for 0.6.x beside the master branch.
Yes please. I will update karafka references immediately after that.

**Comment by bradurani** [comment]
@nguyenquangminh0711 thank you sincerely for this fix! We really appreciate your work on this gem 👏👏👏👏👏👏👏👏

**Comment by dasch** [comment]
I've released v0.7.0.beta2 and v0.6.8 containing the fix for this issue.

**Comment by mainameiz** [comment]
пиздец
